data/core/doc/highlighter.lua

```diff
@@ -26,7 +30,13 @@ function Highlighter:new(doc)
           local state = (i > 1) and self.lines[i - 1].state
           local line = self.lines[i]
           if not (line and line.init_state == state and line.text == self.doc.lines[i]) then
+            local start_time = system.get_time();
             self.lines[i] = self:tokenize_line(i, state)
+            local total_time = system.get_time() - start_time
+            if total_time >= considered_slow then
+              core.redraw = true
+              coroutine.yield(0)
+            end
```

Comment:
What about something like:
```lua
        local start_time = system.get_time();
        for i = self.first_invalid_line, max do
          local state = (i > 1) and self.lines[i - 1].state
          local line = self.lines[i]
          if not (line and line.init_state == state and line.text == self.doc.lines[i]) then
            self.lines[i] = self:tokenize_line(i, state)
            local end_time = system.get_time()
            local total_time = end_time - start_time
            if total_time >= considered_slow then
              core.redraw = true
              coroutine.yield(0)
              start_time = system.get_time();
            else
              start_time = end_time
            end
          end
        end
```
to avoid calling `system.get_time` too much?

Also, is `core.redraw = true` needed?

Reply:
> to avoid calling `system.get_time` too much?

Ahh nice, didn't thought about that one, will change!

> Also, is `core.redraw = true` needed?

I tested without core.redraw and at least on really slow lines I saw the improvement with core.redraw enabled. When drawing takes 0.05s a max of 20 lines can be draw in one second so it would call core.redraw 20 times in a second which I don't think isn't that bad, taking into consideration that it will only do this if each consecutive line takes 0.05 seconds or more to draw. But, maybe I can increase the threshold only for redraw like:

```lua
if total_time > 0.09 then
  core.redraw = true
end
```
Wich would call core.redraw a max of 11 times per second, again, depending on how many lines are slowly tokenized. Will experiment to see.